core/src/dealer/mod.rs

```diff
@@ -0,0 +1,533 @@
+    fn handle_request(&self, request: Request<Payload>, responder: Responder);
+}
+
+type MessageHandler = mpsc::UnboundedSender<Message<JsonValue>>;
+
+// TODO: Maybe it's possible to unregister subscription directly when they
+//       are dropped instead of on next failed attempt.
+pub struct Subscription(UnboundedReceiver<Message<JsonValue>>);
+
+impl Stream for Subscription {
+    type Item = Message<JsonValue>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        self.0.poll_recv(cx)
+    }
+}
+
+#[derive(Default)]
+pub struct Builder {
+    message_handlers: SubscriberMap<MessageHandler>,
+    request_handlers: HandlerMap<Box<dyn RequestHandler>>,
+}
+
+macro_rules! create_dealer {
+    ($builder:expr, $shared:ident -> $body:expr) => {
+        match $builder {
+            builder => {
+                let shared = Arc::new(DealerShared {
+                    message_handlers: Mutex::new(builder.message_handlers),
+                    request_handlers: Mutex::new(builder.request_handlers),
+                    notify_drop: Semaphore::new(0),
+                });
+
+                let handle = {
+                    let $shared = Arc::clone(&shared);
+                    tokio::spawn($body)
+                };
+
+                Dealer {
+                    shared,
+                    handle: Some(handle),
+                }
+            }
+        }
+    };
+}
+
+impl Builder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add_handler<H: RequestHandler + Sized + 'static>(
+        &mut self,
+        uri: &str,
+        handler: H,
+    ) -> Result<(), AlreadyHandledError> {
+        self.request_handlers
+            .insert(split_uri(uri), Box::new(handler))
+    }
+
+    pub fn subscribe(&mut self, uris: &[&str]) -> Subscription {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let handlers = &mut self.message_handlers;
+
+        for &uri in uris {
+            handlers.insert(split_uri(uri), tx.clone());
+        }
+
+        Subscription(rx)
+    }
+
+    pub fn launch_in_background<Fut, F>(self, get_url: F, proxy: Option<Url>) -> Dealer
+    where
+        Fut: Future<Output = Url> + Send + 'static,
+        F: (FnMut() -> Fut) + Send + 'static,
+    {
+        create_dealer!(self, shared -> run(shared, None, get_url, proxy))
+    }
+
+    pub async fn launch<Fut, F>(self, mut get_url: F, proxy: Option<Url>) -> WsResult<Dealer>
+    where
+        Fut: Future<Output = Url> + Send + 'static,
+        F: (FnMut() -> Fut) + Send + 'static,
+    {
+        let dealer = create_dealer!(self, shared -> {
+            // Try to connect.
+            let url = get_url().await;
+            let tasks = connect(&url, proxy.as_ref(), &shared).await?;
+
+            // If a connection is established, continue in a background task.
+            run(shared, Some(tasks), get_url, proxy)
+        });
```

Comment:
I don't use macros very often, but I like this solution. Note that one of the functions is `async` and returns a result, but the other doesn't.